single_kernel_mongo/state/charm_state.py

```diff
@@ -278,33 +306,63 @@ def peer_units_data_interfaces(self) -> dict[Unit, DataPeerOtherUnitData]:
             for unit in self.peers_units
         }
 
+    @property
+    def formatted_socket_path(self) -> str:
+        """URL encoded socket path."""
+        return quote(f"{self.paths.socket_path}", safe="")
+
     @property
     def app_hosts(self) -> set[str]:
         """Retrieve the hosts associated with MongoDB application."""
+        if (
+            self.substrate == Substrates.VM
+            and self.charm_role.name == KindEnum.MONGOS
+            and not self.app_peer_data.external_connectivity
+        ):
+            return {self.formatted_socket_path}
         return {unit.host for unit in self.units}
 
     @property
     def internal_hosts(self) -> set[str]:
```

Comment:
QQ is this function and one above really needed? is it possible to combine them?